### ISSUE TITLE
NPE when Injecting Redirect for Constructor that doesn't exist

### ISSUE THREAD

**Issue by phase** [SpongePowered/Mixin#247]
I've got a weird setup going...
Currently, `MinecraftServer` is calling an invalid constructor for `FrozenThread`:

```java
public class MinecraftServer {
    public MinecraftServer() {
        new FrozenThread(this); // This constructor doesn't exist
    }
}
```

```java
public class FrozenThread {
    public FrozenThread(MinecraftClient client) {
        // this constructor takes a MinecraftClient instead of a MinecraftServer
    }
}
```

I'm attempting to apply this mixin:

```java
@Mixin(MinecraftServer.class)
public abstract class MixinMinecraftServer implements Runnable, ICommandListener {

    @Redirect(method = "<init>", at = @At(value = "NEW", target="(Lnet/minecraft/server/MinecraftServer;)Lnet/minecraft/FrozenThread;"))
    public FrozenThread fixFrozenThread(MinecraftServer minecraftServer) {
        // return valid FrozenThread instance
    }

}
```

When applying, I get this error:

```
[04:27:38] [main/WARN]: Mixin apply failed mixins.pigeon.merge.json:MixinMinecraftServer -> net.minecraft.server.MinecraftServer: org.spongepowered.asm.mixin.transformer.throwables.InvalidMixinException Unexpecteded NullPointerException whilst applying the mixin class: null
org.spongepowered.asm.mixin.transformer.throwables.InvalidMixinException: Unexpecteded NullPointerException whilst applying the mixin class: null
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.apply(MixinApplicatorStandard.java:279) ~[mixin-0.7.7-SNAPSHOT.jar:0.7.7-SNAPSHOT+unknown-b0.git-unknown]
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.applyMixins(TargetClassContext.java:353) ~[mixin-0.7.7-SNAPSHOT.jar:0.7.7-SNAPSHOT+unknown-b0.git-unknown]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.apply(MixinTransformer.java:724) [mixin-0.7.7-SNAPSHOT.jar:0.7.7-SNAPSHOT+unknown-b0.git-unknown]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.applyMixins(MixinTransformer.java:703) [mixin-0.7.7-SNAPSHOT.jar:0.7.7-SNAPSHOT+unknown-b0.git-unknown]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClassBytes(MixinTransformer.java:509) [mixin-0.7.7-SNAPSHOT.jar:0.7.7-SNAPSHOT+unknown-b0.git-unknown]
	at org.spongepowered.asm.mixin.transformer.Proxy.transform(Proxy.java:72) [mixin-0.7.7-SNAPSHOT.jar:0.7.7-SNAPSHOT+unknown-b0.git-unknown]
	at net.minecraft.launchwrapper.LaunchClassLoader.runTransformers(LaunchClassLoader.java:502) [legacylauncher-1.18.jar:?]
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:216) [legacylauncher-1.18.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424) [?:1.8.0_101]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) [?:1.8.0_101]
	at java.lang.Class.forName0(Native Method) [?:1.8.0_101]
	at java.lang.Class.forName(Class.java:348) [?:1.8.0_101]
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:147) [legacylauncher-1.18.jar:?]
	at net.minecraft.launchwrapper.Launch.main(Launch.java:31) [legacylauncher-1.18.jar:?]
	at io.jadon.pigeon.launcher.PigeonLauncher.launchServer(launcher.kt:31) [classes/:?]
	at io.jadon.pigeon.launcher.PigeonLauncher.main(launcher.kt:19) [classes/:?]
Caused by: java.lang.NullPointerException
	at org.spongepowered.asm.mixin.injection.invoke.RedirectInjector.injectAtConstructor(RedirectInjector.java:651) ~[mixin-0.7.7-SNAPSHOT.jar:0.7.7-SNAPSHOT+unknown-b0.git-unknown]
	at org.spongepowered.asm.mixin.injection.invoke.RedirectInjector.inject(RedirectInjector.java:267) ~[mixin-0.7.7-SNAPSHOT.jar:0.7.7-SNAPSHOT+unknown-b0.git-unknown]
	at org.spongepowered.asm.mixin.injection.code.Injector.inject(Injector.java:199) ~[mixin-0.7.7-SNAPSHOT.jar:0.7.7-SNAPSHOT+unknown-b0.git-unknown]
	at org.spongepowered.asm.mixin.injection.struct.InjectionInfo.inject(InjectionInfo.java:267) ~[mixin-0.7.7-SNAPSHOT.jar:0.7.7-SNAPSHOT+unknown-b0.git-unknown]
	at org.spongepowered.asm.mixin.transformer.MixinTargetContext.applyInjections(MixinTargetContext.java:1198) ~[mixin-0.7.7-SNAPSHOT.jar:0.7.7-SNAPSHOT+unknown-b0.git-unknown]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.applyInjections(MixinApplicatorStandard.java:909) ~[mixin-0.7.7-SNAPSHOT.jar:0.7.7-SNAPSHOT+unknown-b0.git-unknown]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.applyMixin(MixinApplicatorStandard.java:309) ~[mixin-0.7.7-SNAPSHOT.jar:0.7.7-SNAPSHOT+unknown-b0.git-unknown]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.apply(MixinApplicatorStandard.java:267) ~[mixin-0.7.7-SNAPSHOT.jar:0.7.7-SNAPSHOT+unknown-b0.git-unknown]
	... 15 more
```

which says that `meta` in

https://github.com/SpongePowered/Mixin/blob/46763a31e50049702b8ccb75d6ee83c6a913b9e2/src/main/java/org/spongepowered/asm/mixin/injection/invoke/RedirectInjector.java#L651

is null.

---

If this is a terrible way to go about doing this, I'm all ears. My main goal here is combing two jars, adding constructors and methods that are expected in classes from one jar to classes shared by both. Adding methods via Mixins is easy enough, but it doesn't support constructors. An ideal solution would be to add a constructor to `FrozenThread` to accept all instances of classes calling the constructor that doesn't exist.

```java
@Mixin(FrozenThread.class)
public abstract class FrozenThread extends Thread {
    public FrozenThread(MinecraftServer server) {
        // do stuff
    }
}
```

**Comment by Mumfrey** [comment]
I'd say the way you're going about it is as sensible as it's possible to be, it's obviously a bit odd that a class is able to be compiled against a class with a ctor which doesn't exist but with respect to fixing the issue I'd say you're using the most sensible approach.

As to why the `meta` decoration is ending up null, that's really weird and definitely a bug, in so far as it either shouldn't get to that state where it's injecting on an undecorated node, or at very least should throw a more useful error if the problem lies elsewhere. It's damn peculiar that it's able to hit the injector logic without the decoration taking place so either this is something wrong with the injector, or potentially some interaction between other ctor handling elsewhere in Mixin which is interacting with the injector in an unpredictable way.

I will investigate and see what I find.

**Comment by Mumfrey** [comment]
> . Adding methods via Mixins is easy enough, but it doesn't support constructors. An ideal solution would be to add a constructor to FrozenThread to accept all instances of classes calling the constructor that doesn't exist.

To elaborate on why this isn't supported in Mixin, there are two reasons:

* Under normal circumstances, there would be no use-case for this scenario because there wouldn't be any way to call the constructor without using reflection anyway. This is the minor reason.
* Constructors as they exist in compiled code are something of a Frankenstein's Monster of mashed-together parts and there is no reliable way to combine them without risking some serious breakage. This is the major reason.

Constructors in the compiled class are a single method which actually combines the following parts of a source Java file into a single method body (not in this order):

* The constructor code as written by the user
* Invocation of default super-constructor if *implied* (eg. no explicit super-ctor call)
* Instance field initialisers as defined against instance fields
* Instance initialiser blocks (non-field)
* Initialisation of synthetic class members, eg. outer class `this$` and other "closure"-type members

This is all packed into a single method named `<init>` and so code which appears once in the Java file may appear more than once in the compiled class if the class has multiple ctors, since initialiser code needs to be packed into each one.

Since the class can end up in an undpredictable state (at least a state which cannot be inferred from reading the original Java code) if initialisers are not run, it is important that all expected initialisers run in the order implied by the contract of the Java language, the class is not considered ready for use until all initialisers, and user ctor code, have completed.

Mixin *ignores* user constructors in mixins which is why you can define anything you like inside them in order to satisfy the compiler and let the class compile. Mixin *merges* field initialisers into all ctors in the target class, but leaves all existing initialisers intact, this is because there is no safe way to identify and then replace relevant initialisers in the target class.

In order to merge a custom ctor therefore, all initialisers in the target class would need to be identified and combined into the merged ctor. This would require careful unpicking of the initialiser code from the original ctor code, which is not reliably achievable. The bytecode of the initialiser is not even necessarily re-usable as-is because the frame of the ctor will include user ctor code from the original Java, as well as initialisation of synthetic members, which might have members in the frame, especially if the ctor itself has arguments.

Basically none of this is completely impossible, but it's a terrifically complex situation to deal with which when coupled with consideration 1 (that it's not a particularly useful function to have 99% of the time) means that supporting it would be adding a lot of complexity for not much utility.

I think there are certainly cases where it would be possible to create a ctor which could be merged, but in the majority of cases it's going to come unstuck quite quickly, so limiting ctor-altering functionality to the current feature set is the most prudent approach, which is why it's set up that way.

**Comment by phase** [comment]
> it's obviously a bit odd that a class is able to be compiled against a class with a ctor which doesn't exist

It wasn't compiled against the version it's trying to call, it was compiled against a different version of the class. These switched classes are classes that appear in both the client & the server (the server only classes are then mapped to the shared mappings in [this file](https://github.com/PigeonMC/BetaMappings/blob/977539b48043292815c53e8e786d35b8406d80fb/serverToClientObf.tsrg)), and the client versions are used at runtime.

Here's a classic _Mumfrey Diagram_ to explain what's going on: <sub>(I don't know what program you use so I used GIMP)</sub>

![very shitty diagram made in gimp](https://i.imgur.com/LBZNpFC.png)

Right now I'm at the step of adding the mixins to fix the merge issues, which includes functions and constructors that the server expects the in shared classes.

I was trying a different solution with constructor injection and quickly ran into the problem of _"wow this is a fuckin mess"_. Hopefully this solution gets sorted out and works properly!

**Comment by Mumfrey** [comment]
> It wasn't compiled against the version it's trying to call, it was compiled against a different version of the class. These switched classes are classes that appear in both the client & the server (the server only classes are then mapped to the shared mappings in this file), and the client versions are used at runtime.

I figured as much, this is a similar approach to Forge but they overcome the issues using patches, which is a maintenance burden.

> Here's a classic Mumfrey Diagram to explain what's going on: (I don't know what program you use so I used GIMP)

Beautiful. For reference I use (and thoroughly recommend) Microsoft Visio for all of my diagrams.

> Right now I'm at the step of adding the mixins to fix the merge issues, which includes functions and constructors that the server expects in the shared classes.
> 
> I was trying a different solution with constructor injection and quickly ran into the problem of "wow this is a fuckin mess". Hopefully this solution gets sorted out and works properly!

Yes this doesn't surprise me in the slightest, constructors are a minefield because of their composite nature as described above. It's certainly nontrivial to unpick them in bytecode form and reconstitute them in any meaningful way. For reasons that should be obvious by now, it's not just a case of lifting the ctor from one object and merging it into the other.

It's certainly not straightforward to actually create a test-case for this, I've been building some MCVE test case classes using *Jasmin* but they aren't triggering the issue. So maybe I can grab your project in order to explore the error condition.

**Comment by phase** [comment]
> I figured as much, this is a similar approach to Forge but they overcome the issues using patches, which is a maintenance burden.

I tried [patches](https://github.com/phase/gamma/tree/master/patches/minecraft/unknown) before, but it quickly became a hassle.

> So maybe I can grab your project in order to explore the error condition.

erhh, you can [_attempt_](https://github.com/PigeonMC/Pigeon#cloning) to do that. (Don't ask why I'm modding Beta 1.7.3, Zidane already laughed at me for it.) I think it should build out of the box, but I don't have a `gradle run` yet because you need to add the LWJGL Natives to the VM arguments and I just haven't gotten around to doing it.

So to run it:

* `git clone https://github.com/PigeonMC/Pigeon.git --recursive`
* `./gradlew build` (this should work, probably)
  * It will download the Beta 1.7.3 client fine, but it won't get the Beta 1.7.3 server because Mojang doesn't distribute it anymore. You can download the server from here (direct download): [https://jadon.io/files/minecraft_server.jar](https://jadon.io/files/minecraft_server.jar).
  * I also have uploaded a finalized jar that contains all the modifications I'm making to it. If this process fails, you can download it here: [https://jadon.io/files/minecraft_merged_mapped.jar](https://jadon.io/files/minecraft_merged_mapped.jar)
* Open the `build.gradle` in IntelliJ
* Download the [LWJGL Natives](http://legacy.lwjgl.org/download.php.html) and put it the root `Pigeon/`
* Add `-Djava.library.path=native/linux/` to the VM arguments
* Run `io.pigeon.launcher.Pigeon` to run. You might need to change `launchClient` to `launchServer` in `main`.

There are a lot of [hacky](https://github.com/PigeonMC/Pigeon/blob/159a60fccade1d1822d632f55e261d616e9be57b/buildSrc/src/main/kotlin/io/jadon/pigeon/jar.kt#L47-L60) things going on, but everything should work as normal. (The client launches perfectly :tada:)

**Comment by Mumfrey** [comment]
Ok I finally got your project to work. I had to disable the 'mapMinecraft` task because the jar it generated was useless, I used the merged jar you linked instead. However getting it to run was a real nuisance. For reference you have put java classes in your `src/main/kotlin` dir instead of `src/main/java` so they don't get compiled!

Once I moved the java files to the correct location and built a shaded jar I was able to run the project and reproduce the error, so I can begin looking into it now.

**Comment by Mumfrey** [comment]
So I found the problem, this is actually a bug in my handling of *wildcard* ctor redirects. For wildcard redirects the meta is created and for non-wildcard the meta is not used and the absence of the meta was used as the semaphore for this state. However I wasn't taking that into account in the injector body itself.

I have changed the injector to always add the decoration and store the `wildcard` semaphore in the metadata object, this means the meta is always present and the other wildcard decoration can go away.
